addons/mass_mailing/models/mailing.py

```diff
@@ -472,6 +478,7 @@ def create(self, vals_list):
     def write(self, values):
         if values.get('body_html'):
             values['body_html'] = self._convert_inline_images_to_urls(values['body_html'])
+            values['body_html'] = self._crop_vml_images(values['body_html'])
```

Comment:
For the moment there are two image adaptations call to make. If we ever do more then I think it might be worth having an indirection with a more generic function like `_adaptImages` that would call all those adaptations because otherwise it would be too easy to forget to call one of the bunch. For now I think it's probably overkill but it's something to keep in mind for the future.

Reply:
That's a good idea. I could do something like this:
``` diff
diff --git a/addons/mass_mailing/models/mailing.py b/addons/mass_mailing/models/mailing.py
index 978e6059b39..9a731da4a79 100644
--- a/addons/mass_mailing/models/mailing.py
+++ b/addons/mass_mailing/models/mailing.py
@@ -468,8 +468,11 @@ class MassMailing(models.Model):
         ab_testing_cron = self.env.ref('mass_mailing.ir_cron_mass_mailing_ab_testing').sudo()
         for values in vals_list:
             if values.get('body_html'):
-                values['body_html'] = self._convert_inline_images_to_urls(values['body_html'])
-                values['body_html'] = self._crop_vml_images(values['body_html'])
+                parsed_body = lxml.html.fromstring(values['body_html'])
+                modified = self._convert_inline_images_to_urls(parsed_body) or modified
+                modified = self._crop_vml_images(parsed_body) or modified
+                if modified:
+                    values['body_html'] = lxml.html.tostring(parsed_body, encoding='unicode')
             if values.get('ab_testing_schedule_datetime'):
                 at = fields.Datetime.from_string(values['ab_testing_schedule_datetime'])
                 ab_testing_cron._trigger(at=at)
@@ -481,8 +485,12 @@ class MassMailing(models.Model):
 
     def write(self, values):
         if values.get('body_html'):
-            values['body_html'] = self._convert_inline_images_to_urls(values['body_html'])
-            values['body_html'] = self._crop_vml_images(values['body_html'])
+            parsed_body = lxml.html.fromstring(values['body_html'])
+            modified = self._convert_inline_images_to_urls(parsed_body) or modified
+            modified = self._crop_vml_images(parsed_body) or modified
+            if modified:
+                values['body_html'] = lxml.html.tostring(parsed_body, encoding='unicode')
         # If ab_testing is already enabled on a mailing and the campaign is removed, we raise a ValidationError
         if values.get('campaign_id') is False and any(mailing.ab_testing_enabled for mailing in self) and 'ab_testing_enabled' not in values:
             raise ValidationError(_("A campaign should be set when A/B test is enabled"))
@@ -1367,12 +1375,11 @@ class MassMailing(models.Model):
                 )
             )
 
-    def _convert_inline_images_to_urls(self, body_html):
+    def _convert_inline_images_to_urls(self, root):
         """
         Find inline base64 encoded images, make an attachement out of
         them and replace the inline image with an url to the attachement.
         """
-        root = lxml.html.fromstring(body_html)
         modified = False
         for node in root.iter('img'):
             match = image_re.match(node.attrib.get('src', ''))
@@ -1389,17 +1396,13 @@ class MassMailing(models.Model):
                 image = re.sub(r'data:image/[A-Za-z]+;base64,', '', match).encode()  # base64 image as bytes
                 comment.text = comment.text.replace(match, self._image_to_url(image))
 
-        if modified:
-            return lxml.html.tostring(root, encoding='unicode')
+        return modified
 
-        return body_html
-
-    def _crop_vml_images(self, body_html):
+    def _crop_vml_images(self, root):
         """
         Find VML v:image elements, crop their source images, make an attachement
         out of them and replace their source with an url to the attachement.
         """
-        root = lxml.html.fromstring(body_html)
         modified = False
         with requests.Session() as session:
             for comment in [c for c in root.iter(lxml.etree.Comment) if mso_re.match(c.text)]:
@@ -1427,7 +1430,5 @@ class MassMailing(models.Model):
                         # the entire process just for one image's responsive
                         # cropping behavior).
                         pass
-        if modified:
-            return lxml.html.tostring(root, encoding='unicode')
 
-        return body_html
+        return modified
```
Now that's 5 lines instead of 2 that are duplicated  between `create` and `write` which is no big deal but I could still create a separate function to group that together, like `_handle_mail_body_images`...? cc @dmo-odoo